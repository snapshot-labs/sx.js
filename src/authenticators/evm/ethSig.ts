import EthSigAuthenticatorAbi from './abis/EthSigAuthenticator.json';
import { getRSVFromSig, hexPadLeft } from '../../utils/encoding';
import type {
  Authenticator,
  Envelope,
  Propose,
  UpdateProposal,
  Vote,
  Call
} from '../../clients/evm/types';

export default function createEthSigAuthenticator(): Authenticator {
  return {
    type: 'ethSig',
    createCall(
      envelope: Envelope<Propose | UpdateProposal | Vote>,
      selector: string,
      calldata: string[]
    ): Call {
      const { signatureData, data } = envelope;
      const { space } = data;

      if (!signatureData) throw new Error('signatureData is required for this authenticator');

      const { r, s, v } = getRSVFromSig(signatureData.signature);

      const args = [
        v,
        hexPadLeft(r.toHex()),
        hexPadLeft(s.toHex()),
        signatureData.message.salt || '0x00',
        space,
        selector,
        ...calldata
      ];

      return {
        abi: EthSigAuthenticatorAbi,
        args
      };
    }
  };
}
