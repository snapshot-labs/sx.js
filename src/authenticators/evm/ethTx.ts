import EthTxAuthenticatorAbi from './abis/EthTxAuthenticator.json';
import type {
  Authenticator,
  Envelope,
  Propose,
  UpdateProposal,
  Vote,
  Call
} from '../../clients/evm/types';

export default function createEthTxAuthenticator(): Authenticator {
  return {
    type: 'ethTx',
    createCall(
      envelope: Envelope<Propose | UpdateProposal | Vote>,
      selector: string,
      calldata: string[]
    ): Call {
      const { space } = envelope.data;

      return {
        abi: EthTxAuthenticatorAbi,
        args: [space, selector, ...calldata]
      };
    }
  };
}
