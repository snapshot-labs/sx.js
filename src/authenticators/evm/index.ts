import createVanillaAuthenticator from './vanilla';
import createEthTxAuthenticator from './ethTx';
import createEthSigAuthenticator from './ethSig';
import type { Authenticator } from '../../clients/evm/types';
import type { EvmNetworkConfig } from '../../types';

export function getAuthenticator(
  address: string,
  networkConfig: EvmNetworkConfig
): Authenticator | null {
  const authenticator = networkConfig.authenticators[address];
  if (!authenticator) return null;

  if (authenticator.type === 'vanilla') {
    return createVanillaAuthenticator();
  }

  if (authenticator.type === 'ethTx') {
    return createEthTxAuthenticator();
  }

  if (authenticator.type === 'ethSig') {
    return createEthSigAuthenticator();
  }

  return null;
}
