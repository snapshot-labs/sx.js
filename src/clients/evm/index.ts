import { Contract } from '@ethersproject/contracts';
import { Interface } from '@ethersproject/abi';
import randomBytes from 'randombytes';
import VanillaAuthenticatorAbi from './abis/VanillaAuthenticator.json';
import SpaceAbi from './abis/Space.json';
import SpaceFactoryAbi from './abis/SpaceFactory.json';
import type { Signer } from '@ethersproject/abstract-signer';

type Choice = 0 | 1 | 2;

type AddressConfig = {
  addy: string;
  params: string;
};

type IndexedConfig = {
  index: number;
  params: string;
};

export class SnapshotEVMClient {
  async deploy({
    signer,
    spaceFactory,
    owner,
    votingDelay,
    minVotingDuration,
    maxVotingDuration,
    proposalThreshold,
    quorum,
    authenticators,
    votingStrategies,
    executionStrategies
  }: {
    signer: Signer;
    spaceFactory: string;
    owner: string;
    votingDelay: number;
    minVotingDuration: number;
    maxVotingDuration: number;
    proposalThreshold: number;
    quorum: number;
    authenticators: string[];
    votingStrategies: AddressConfig[];
    executionStrategies: string[];
  }): Promise<{ txId: string; spaceAddress: string }> {
    const spaceFactoryContract = new Contract(spaceFactory, SpaceFactoryAbi, signer);

    const salt = `0x${randomBytes(32).toString('hex')}`;
    const args = [
      owner,
      votingDelay,
      minVotingDuration,
      maxVotingDuration,
      proposalThreshold,
      quorum,
      votingStrategies,
      authenticators,
      executionStrategies,
      salt
    ];

    const [spaceAddress, response] = await Promise.all([
      spaceFactoryContract.getSpaceAddress(...args),
      spaceFactoryContract.createSpace(...args)
    ]);

    return { spaceAddress, txId: response.hash };
  }

  async propose({
    signer,
    space,
    authenticator,
    userVotingStrategies,
    executionStrategy,
    metadataUri
  }: {
    signer: Signer;
    space: string;
    authenticator: string;
    userVotingStrategies: IndexedConfig[];
    executionStrategy: AddressConfig;
    metadataUri: string;
  }) {
    const spaceInterface = new Interface(SpaceAbi);
    const authenticatorContract = new Contract(authenticator, VanillaAuthenticatorAbi, signer);

    const proposerAddress = await signer.getAddress();

    const functionData = spaceInterface.encodeFunctionData('propose', [
      proposerAddress,
      metadataUri,
      executionStrategy,
      userVotingStrategies
    ]);

    const selector = functionData.slice(0, 10);
    const calldata = `0x${functionData.slice(10)}`;

    return authenticatorContract.authenticate(space, selector, calldata);
  }

  async vote({
    signer,
    space,
    authenticator,
    userVotingStrategies,
    proposal,
    choice
  }: {
    signer: Signer;
    space: string;
    authenticator: string;
    userVotingStrategies: IndexedConfig[];
    proposal: number;
    choice: Choice;
  }) {
    const spaceInterface = new Interface(SpaceAbi);
    const authenticatorContract = new Contract(authenticator, VanillaAuthenticatorAbi, signer);

    const voterAddress = await signer.getAddress();

    const functionData = spaceInterface.encodeFunctionData('vote', [
      voterAddress,
      proposal,
      choice,
      userVotingStrategies
    ]);

    const selector = functionData.slice(0, 10);
    const calldata = `0x${functionData.slice(10)}`;

    return authenticatorContract.authenticate(space, selector, calldata);
  }

  async finalizeProposal({
    signer,
    space,
    proposal,
    executionParams
  }: {
    signer: Signer;
    space: string;
    proposal: number;
    executionParams: string;
  }) {
    const spaceContract = new Contract(space, SpaceAbi, signer);

    return spaceContract.finalizeProposal(proposal, executionParams);
  }

  async cancelProposal({
    signer,
    space,
    proposal,
    executionParams
  }: {
    signer: Signer;
    space: string;
    proposal: number;
    executionParams: string;
  }) {
    const spaceContract = new Contract(space, SpaceAbi, signer);

    return spaceContract.cancelProposal(proposal, executionParams);
  }

  async getProposal({
    signer,
    space,
    proposal
  }: {
    signer: Signer;
    space: string;
    proposal: number;
  }) {
    const spaceContract = new Contract(space, SpaceAbi, signer);

    return spaceContract.getProposal(proposal);
  }

  async getProposalStatus({
    signer,
    space,
    proposal
  }: {
    signer: Signer;
    space: string;
    proposal: number;
  }) {
    const spaceContract = new Contract(space, SpaceAbi, signer);

    return spaceContract.getProposalStatus(proposal);
  }

  async setMaxVotingDuration({
    signer,
    space,
    maxVotingDuration
  }: {
    signer: Signer;
    space: string;
    maxVotingDuration: number;
  }) {
    const spaceContract = new Contract(space, SpaceAbi, signer);

    return spaceContract.setMaxVotingDuration(maxVotingDuration);
  }

  async setMinVotingDuration({
    signer,
    space,
    minVotingDuration
  }: {
    signer: Signer;
    space: string;
    minVotingDuration: number;
  }) {
    const spaceContract = new Contract(space, SpaceAbi, signer);

    return spaceContract.setMinVotingDuration(minVotingDuration);
  }

  async setMetadataUri({
    signer,
    space,
    metadataUri
  }: {
    signer: Signer;
    space: string;
    metadataUri: string;
  }) {
    const spaceContract = new Contract(space, SpaceAbi, signer);

    return spaceContract.setMetadataUri(metadataUri);
  }

  async setProposalThreshold({
    signer,
    space,
    threshold
  }: {
    signer: Signer;
    space: string;
    threshold: number;
  }) {
    const spaceContract = new Contract(space, SpaceAbi, signer);

    return spaceContract.setProposalThreshold(threshold);
  }

  async setVotingDelay({
    signer,
    space,
    votingDelay
  }: {
    signer: Signer;
    space: string;
    votingDelay: number;
  }) {
    const spaceContract = new Contract(space, SpaceAbi, signer);

    return spaceContract.setVotingDelay(votingDelay);
  }
}
