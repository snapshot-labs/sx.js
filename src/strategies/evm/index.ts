import createVanillaStrategy from './vanilla';
import createCompStrategy from './comp';
import createOzVotesStrategy from './ozVotes';
import createMerkleWhitelist from './merkleWhitelist';
import type {
  Propose,
  Vote,
  StrategyConfig,
  Strategy,
  IndexedConfig
} from '../../clients/evm/types';
import type { EvmNetworkConfig } from '../../types';

export function getStrategy(address: string, networkConfig: EvmNetworkConfig): Strategy | null {
  const strategy = networkConfig.strategies[address];
  if (!strategy) return null;

  if (strategy.type === 'vanilla') {
    return createVanillaStrategy();
  }

  if (strategy.type === 'comp') {
    return createCompStrategy();
  }

  if (strategy.type === 'ozVotes') {
    return createOzVotesStrategy();
  }

  if (strategy.type === 'whitelist') {
    return createMerkleWhitelist();
  }

  return null;
}

export async function getStrategiesWithParams(
  call: 'propose' | 'vote',
  strategies: StrategyConfig[],
  signerAddress: string,
  data: Propose | Vote,
  networkConfig: EvmNetworkConfig
) {
  const results = await Promise.all(
    strategies.map(async strategyConfig => {
      const strategy = getStrategy(strategyConfig.address, networkConfig);
      if (!strategy) throw new Error('Invalid strategy');

      try {
        const params = await strategy.getParams(
          call,
          strategyConfig,
          signerAddress,
          strategyConfig.metadata || null,
          data
        );

        return {
          index: strategyConfig.index,
          params
        };
      } catch (e) {
        return null;
      }
    })
  );

  return results.filter(Boolean) as IndexedConfig[];
}
