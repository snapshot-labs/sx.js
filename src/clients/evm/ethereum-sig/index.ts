import randomBytes from 'randombytes';
import { AbiCoder } from '@ethersproject/abi';
import { SplitUint256 } from '../../../utils/split-uint256';
import { bytesToHex } from '../../../utils/bytes';
import { getStrategiesWithParams } from '../../../strategies/evm';
import { evmGoerli } from '../../../networks';
import { domain as baseDomain, proposeTypes, updateProposalTypes, voteTypes } from './types';
import type { Signer, TypedDataSigner, TypedDataField } from '@ethersproject/abstract-signer';
import type {
  Propose,
  UpdateProposal,
  Vote,
  Envelope,
  SignatureData,
  EIP712ProposeMessage,
  EIP712UpdateProposalMessage,
  EIP712VoteMessage
} from '../types';
import type { EvmNetworkConfig } from '../../../types';

type EthereumSigClientOpts = {
  networkConfig?: EvmNetworkConfig;
  manaUrl?: string;
};

export class EthereumSig {
  manaUrl: string;
  networkConfig: EvmNetworkConfig;

  constructor(opts?: EthereumSigClientOpts) {
    this.networkConfig = opts?.networkConfig || evmGoerli;
    this.manaUrl = opts?.manaUrl || 'https://mana.pizza';
  }

  generateSalt() {
    return Number(SplitUint256.fromHex(bytesToHex(randomBytes(4))).toHex());
  }

  public async sign<
    T extends EIP712ProposeMessage | EIP712UpdateProposalMessage | EIP712VoteMessage
  >(
    signer: Signer & TypedDataSigner,
    verifyingContract: string,
    message: T,
    types: Record<string, TypedDataField[]>
  ): Promise<SignatureData> {
    const address = await signer.getAddress();

    const domain = {
      ...baseDomain,
      chainId: this.networkConfig.eip712ChainId,
      verifyingContract
    };

    const signature = await signer._signTypedData(domain, types, message);

    return {
      address,
      signature,
      domain,
      types,
      message
    };
  }

  public async send(envelope) {
    const body = {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method: 'send',
        params: { envelope },
        id: null
      })
    };

    const res = await fetch(`${this.manaUrl}/eth_rpc/${this.networkConfig.eip712ChainId}`, body);
    const json = await res.json();

    return json.result;
  }

  public async propose({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: Propose;
  }): Promise<Envelope<Propose>> {
    const author = await signer.getAddress();

    const userStrategies = await getStrategiesWithParams(
      'propose',
      data.strategies,
      author,
      data,
      this.networkConfig
    );

    const abiCoder = new AbiCoder();
    const message: EIP712ProposeMessage = {
      space: data.space,
      author,
      metadataURI: data.metadataUri,
      executionStrategy: data.executionStrategy,
      userProposalValidationParams: abiCoder.encode(
        ['tuple(int8 index, bytes params)[]'],
        [userStrategies]
      ),
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(signer, data.authenticator, message, proposeTypes);

    return {
      signatureData,
      data
    };
  }

  public async updateProposal({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UpdateProposal;
  }): Promise<Envelope<UpdateProposal>> {
    const author = await signer.getAddress();

    const message: EIP712UpdateProposalMessage = {
      space: data.space,
      author,
      proposalId: data.proposal,
      executionStrategy: data.executionStrategy,
      metadataURI: data.metadataUri,
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(signer, data.authenticator, message, updateProposalTypes);

    return {
      signatureData,
      data
    };
  }

  public async vote({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: Vote;
  }): Promise<Envelope<Vote>> {
    const voter = await signer.getAddress();

    const userVotingStrategies = await getStrategiesWithParams(
      'vote',
      data.strategies,
      voter,
      data,
      this.networkConfig
    );

    const message: EIP712VoteMessage = {
      space: data.space,
      voter,
      proposalId: data.proposal,
      choice: data.choice,
      userVotingStrategies,
      voteMetadataURI: data.metadataUri
    };

    const signatureData = await this.sign(signer, data.authenticator, message, voteTypes);

    return {
      signatureData,
      data
    };
  }
}
