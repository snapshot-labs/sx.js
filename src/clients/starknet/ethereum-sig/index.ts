import randomBytes from 'randombytes';
import { Signer, TypedDataSigner, TypedDataField } from '@ethersproject/abstract-signer';
import { CallData, shortString } from 'starknet';
import { getStrategiesWithParams } from '../../../utils/strategies';
import { proposeTypes, updateProposalTypes, voteTypes } from './types';
import {
  ClientConfig,
  ClientOpts,
  Envelope,
  Propose,
  UpdateProposal,
  Vote,
  EIP712ProposeMessage,
  EIP712UpdateProposalMessage,
  EIP712VoteMessage,
  SignatureData
} from '../../../types';
import { defaultNetwork } from '../../..';
import { getRSVFromSig } from '../../../utils/encoding';

export class EthereumSig {
  config: ClientConfig;

  constructor(opts: ClientOpts) {
    this.config = {
      networkConfig: defaultNetwork,
      ...opts
    };
  }

  generateSalt() {
    return `0x${randomBytes(4).toString('hex')}`;
  }

  public async sign<
    T extends EIP712ProposeMessage | EIP712UpdateProposalMessage | EIP712VoteMessage
  >(
    signer: Signer & TypedDataSigner,
    message: T,
    types: Record<string, TypedDataField[]>,
    primaryType: string
  ): Promise<SignatureData> {
    const address = await signer.getAddress();

    const domain = {};

    const extendedMessage = {
      chainId: this.config.networkConfig.eip712ChainId,
      ...message
    };

    const signature = await signer._signTypedData(domain, types, extendedMessage);
    const { r, s, v } = getRSVFromSig(signature);

    return {
      address,
      signature: [r.toHex(), s.toHex(), v],
      message: extendedMessage,
      primaryType
    };
  }

  public async propose({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: Propose;
  }): Promise<Envelope<Propose>> {
    const address = await signer.getAddress();

    const userStrategies = await getStrategiesWithParams(
      'propose',
      data.strategies,
      address,
      data,
      this.config
    );

    const message = {
      authenticator: data.authenticator,
      space: data.space,
      author: address,
      executionStrategy: {
        address: data.executionStrategy.addr,
        params: data.executionStrategy.params
      },
      userProposalValidationParams: CallData.compile({
        userStrategies
      }),
      metadataUri: shortString
        .splitLongString(data.metadataUri)
        .map(str => shortString.encodeShortString(str)),
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(signer, message, proposeTypes, 'Propose');

    return {
      signatureData,
      data
    };
  }

  public async updateProposal({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: UpdateProposal;
  }): Promise<Envelope<UpdateProposal>> {
    const address = await signer.getAddress();

    const message = {
      authenticator: data.authenticator,
      space: data.space,
      author: address,
      proposalId: `0x${data.proposal.toString(16)}`,
      executionStrategy: {
        address: data.executionStrategy.addr,
        params: data.executionStrategy.params
      },
      metadataUri: shortString
        .splitLongString(data.metadataUri)
        .map(str => shortString.encodeShortString(str)),
      salt: this.generateSalt()
    };

    const signatureData = await this.sign(signer, message, updateProposalTypes, 'UpdateProposal');

    return {
      signatureData,
      data
    };
  }

  public async vote({
    signer,
    data
  }: {
    signer: Signer & TypedDataSigner;
    data: Vote;
  }): Promise<Envelope<Vote>> {
    const address = await signer.getAddress();

    const userVotingStrategies = await getStrategiesWithParams(
      'vote',
      data.strategies,
      address,
      data,
      this.config
    );

    const message = {
      authenticator: data.authenticator,
      space: data.space,
      voter: address,
      proposalId: `0x${data.proposal.toString(16)}`,
      choice: `0x${data.choice.toString(16)}`,
      userVotingStrategies,
      metadataUri: shortString.splitLongString('').map(str => shortString.encodeShortString(str))
    };

    const signatureData = await this.sign(signer, message, voteTypes, 'Vote');

    return {
      signatureData,
      data
    };
  }
}
