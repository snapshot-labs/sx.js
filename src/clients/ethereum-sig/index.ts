import { hash, defaultProvider } from 'starknet';
import randomBytes from 'randombytes';
import { Web3Provider } from '@ethersproject/providers';
import { Wallet } from '@ethersproject/wallet';
import fetch from 'cross-fetch';
import { getStrategy } from '../../strategies';
import { domain, proposeTypes, voteTypes } from './types';
import * as utils from '../../utils';
import {
  Propose,
  Vote,
  EthSigProposeMessage,
  EthSigVoteMessage,
  Envelope,
  EthereumSigClientConfig
} from '../../types';

export class EthereumSig {
  config: EthereumSigClientConfig;

  constructor(config: EthereumSigClientConfig) {
    this.config = config;
  }

  generateSalt() {
    return Number(
      utils.splitUint256.SplitUint256.fromHex(utils.bytes.bytesToHex(randomBytes(4))).toHex()
    );
  }

  async getStrategiesAddresses(data: Propose | Vote) {
    return Promise.all(
      data.strategies.map(
        (id) =>
          defaultProvider.getStorageAt(
            data.space,
            utils.encoding.getStorageVarAddress('Voting_voting_strategies_store', id.toString(16))
          ) as Promise<string>
      )
    );
  }

  async getStrategiesParams(call: 'propose' | 'vote', address: string, data: Propose | Vote) {
    const strategiesAddresses = await this.getStrategiesAddresses(data);

    return Promise.all(
      strategiesAddresses.map((strategyAddress, index) => {
        const strategy = getStrategy(strategyAddress);
        if (!strategy) throw new Error('Invalid strategy');

        return strategy.getParams(
          call,
          strategyAddress,
          index,
          {
            address,
            sig: '',
            data: {
              message: data
            }
          },
          this.config
        );
      })
    );
  }

  public async sign<T extends EthSigProposeMessage | EthSigVoteMessage>(
    web3: Web3Provider | Wallet,
    address: string,
    message: T,
    types
  ): Promise<Envelope<T>> {
    const signer = Wallet.isSigner(web3) ? web3 : web3.getSigner();
    const data = { domain, types, message };
    const sig = await signer._signTypedData(domain, data.types, message);
    return { address, sig, data } as any;
  }

  public async send(envelop) {
    const init = {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method: 'send',
        params: { envelop },
        id: null
      })
    };
    const res = await fetch(this.config.manaUrl, init);
    const json = await res.json();
    return json.result;
  }

  public async propose(web3: Web3Provider | Wallet, address: string, data: Propose) {
    const strategiesParams = await this.getStrategiesParams('propose', address, data);

    const message: EthSigProposeMessage = {
      ...data,
      space: utils.encoding.hexPadRight(data.space),
      authenticator: utils.encoding.hexPadRight(data.authenticator),
      proposerAddress: utils.encoding.hexPadRight(address),
      executor: utils.encoding.hexPadRight(data.executor),
      executionParamsHash: utils.encoding.hexPadRight(
        hash.computeHashOnElements(data.executionParams)
      ),
      usedVotingStrategiesHash: utils.encoding.hexPadRight(
        hash.computeHashOnElements(data.strategies.map((strategy) => `0x${strategy.toString(16)}`))
      ),
      userVotingStrategyParamsFlatHash: utils.encoding.hexPadRight(
        hash.computeHashOnElements(utils.encoding.flatten2DArray(strategiesParams))
      ),
      salt: this.generateSalt()
    };

    return this.sign(web3, address, message, proposeTypes);
  }

  public async vote(web3: Web3Provider | Wallet, address: string, data: Vote) {
    const strategiesParams = await this.getStrategiesParams('vote', address, data);

    const message: EthSigVoteMessage = {
      ...data,
      space: utils.encoding.hexPadRight(data.space),
      authenticator: utils.encoding.hexPadRight(data.authenticator),
      voterAddress: utils.encoding.hexPadRight(address),
      usedVotingStrategiesHash: utils.encoding.hexPadRight(
        hash.computeHashOnElements(data.strategies.map((strategy) => `0x${strategy.toString(16)}`))
      ),
      userVotingStrategyParamsFlatHash: utils.encoding.hexPadRight(
        hash.computeHashOnElements(utils.encoding.flatten2DArray(strategiesParams))
      ),
      salt: this.generateSalt()
    };

    return this.sign(web3, address, message, voteTypes);
  }
}
