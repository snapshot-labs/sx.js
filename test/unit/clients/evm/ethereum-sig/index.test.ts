import { JsonRpcProvider } from '@ethersproject/providers';
import { Wallet } from '@ethersproject/wallet';
import { EthereumSig } from '../../../../../src/clients/evm/ethereum-sig';

describe('EthereumSig', () => {
  const space = '0xef8cd9081d7969c5cdcbb84dd8d577d2daefb649';
  const authenticator = '0xc537d997ddc783e071f82ccbfaa0d768d310001b';
  const executor = '0x6241b5c89350bb3c465179706cf26050ea32444f';

  const provider = new JsonRpcProvider('http://127.0.0.1:8545');
  const signer = new Wallet(
    '0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80',
    provider
  );
  const ethSigClient = new EthereumSig();

  beforeAll(() => {
    jest.spyOn(ethSigClient, 'generateSalt').mockImplementation(() => 0);
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should create propose envelope', async () => {
    const envelope = await ethSigClient.propose({
      signer,
      data: {
        space,
        authenticator,
        strategies: [{ index: 0, address: '0xc1245c5dca7885c73e32294140f1e5d30688c202' }],
        executionStrategy: { addr: executor, params: '0x00' },
        metadataUri: 'ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca'
      }
    });

    expect(envelope).toMatchSnapshot();
  });

  it('should create update proposal envelope', async () => {
    const envelope = await ethSigClient.updateProposal({
      signer,
      data: {
        space,
        proposal: 1,
        authenticator,
        executionStrategy: { addr: executor, params: '0x00' },
        metadataUri: 'ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca'
      }
    });

    expect(envelope).toMatchSnapshot();
  });

  it('should create vote envelope', async () => {
    const envelope = await ethSigClient.vote({
      signer,
      data: {
        space,
        authenticator,
        strategies: [{ index: 0, address: '0xc1245c5dca7885c73e32294140f1e5d30688c202' }],
        proposal: 1,
        choice: 1,
        metadataUri: ''
      }
    });

    expect(envelope).toMatchSnapshot();
  });
});
