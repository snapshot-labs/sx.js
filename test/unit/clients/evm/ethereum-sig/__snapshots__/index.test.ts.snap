// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EthereumSig should create propose envelope 1`] = `
{
  "data": {
    "authenticator": "0xc537d997ddc783e071f82ccbfaa0d768d310001b",
    "executionStrategy": {
      "addr": "0x6241b5c89350bb3c465179706cf26050ea32444f",
      "params": "0x00",
    },
    "metadataUri": "ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca",
    "space": "0xef8cd9081d7969c5cdcbb84dd8d577d2daefb649",
    "strategies": [
      {
        "address": "0xc1245c5dca7885c73e32294140f1e5d30688c202",
        "index": 0,
      },
    ],
  },
  "signatureData": {
    "address": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
    "domain": {
      "chainId": 5,
      "name": "snapshot-x",
      "verifyingContract": "0xc537d997ddc783e071f82ccbfaa0d768d310001b",
      "version": "1.0.0",
    },
    "message": {
      "author": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
      "executionStrategy": {
        "addr": "0x6241b5c89350bb3c465179706cf26050ea32444f",
        "params": "0x00",
      },
      "metadataURI": "ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca",
      "salt": 0,
      "space": "0xef8cd9081d7969c5cdcbb84dd8d577d2daefb649",
      "userProposalValidationParams": "0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000",
    },
    "signature": "0x8802f584d5c929c781ea06deefe9d4e9d91dd30386ef8b250850df76044111e167fabb5e5e58ccfffa15ec93f586bfb0acfddbdf6a89d16ba4ede94770eb1f561c",
    "types": {
      "Propose": [
        {
          "name": "space",
          "type": "address",
        },
        {
          "name": "author",
          "type": "address",
        },
        {
          "name": "metadataURI",
          "type": "string",
        },
        {
          "name": "executionStrategy",
          "type": "Strategy",
        },
        {
          "name": "userProposalValidationParams",
          "type": "bytes",
        },
        {
          "name": "salt",
          "type": "uint256",
        },
      ],
      "Strategy": [
        {
          "name": "addr",
          "type": "address",
        },
        {
          "name": "params",
          "type": "bytes",
        },
      ],
    },
  },
}
`;

exports[`EthereumSig should create update proposal envelope 1`] = `
{
  "data": {
    "authenticator": "0xc537d997ddc783e071f82ccbfaa0d768d310001b",
    "executionStrategy": {
      "addr": "0x6241b5c89350bb3c465179706cf26050ea32444f",
      "params": "0x00",
    },
    "metadataUri": "ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca",
    "proposal": 1,
    "space": "0xef8cd9081d7969c5cdcbb84dd8d577d2daefb649",
  },
  "signatureData": {
    "address": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
    "domain": {
      "chainId": 5,
      "name": "snapshot-x",
      "verifyingContract": "0xc537d997ddc783e071f82ccbfaa0d768d310001b",
      "version": "1.0.0",
    },
    "message": {
      "author": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
      "executionStrategy": {
        "addr": "0x6241b5c89350bb3c465179706cf26050ea32444f",
        "params": "0x00",
      },
      "metadataURI": "ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca",
      "proposalId": 1,
      "salt": 0,
      "space": "0xef8cd9081d7969c5cdcbb84dd8d577d2daefb649",
    },
    "signature": "0xea79e53b3ccc11f5ef14883adda7e6858ee2c3fe9823989c8eddab648ec2916b306448cc9de997fd2fb9028a534fe2508cdf5809db0a02a6ed70dcf205bc6c001b",
    "types": {
      "Strategy": [
        {
          "name": "addr",
          "type": "address",
        },
        {
          "name": "params",
          "type": "bytes",
        },
      ],
      "updateProposal": [
        {
          "name": "space",
          "type": "address",
        },
        {
          "name": "author",
          "type": "address",
        },
        {
          "name": "proposalId",
          "type": "uint256",
        },
        {
          "name": "executionStrategy",
          "type": "Strategy",
        },
        {
          "name": "metadataURI",
          "type": "string",
        },
        {
          "name": "salt",
          "type": "uint256",
        },
      ],
    },
  },
}
`;

exports[`EthereumSig should create vote envelope 1`] = `
{
  "data": {
    "authenticator": "0xc537d997ddc783e071f82ccbfaa0d768d310001b",
    "choice": 1,
    "metadataUri": "",
    "proposal": 1,
    "space": "0xef8cd9081d7969c5cdcbb84dd8d577d2daefb649",
    "strategies": [
      {
        "address": "0xc1245c5dca7885c73e32294140f1e5d30688c202",
        "index": 0,
      },
    ],
  },
  "signatureData": {
    "address": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
    "domain": {
      "chainId": 5,
      "name": "snapshot-x",
      "verifyingContract": "0xc537d997ddc783e071f82ccbfaa0d768d310001b",
      "version": "1.0.0",
    },
    "message": {
      "choice": 1,
      "proposalId": 1,
      "space": "0xef8cd9081d7969c5cdcbb84dd8d577d2daefb649",
      "userVotingStrategies": [
        {
          "index": 0,
          "params": "0x00",
        },
      ],
      "voteMetadataURI": "",
      "voter": "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
    },
    "signature": "0xc59afc1cad9abae8416f9b62e2fb51b204a121fbddce31d88e52765b97e307bb22a2f6c396e16da72df77e66b6f1a0052bc2a403c9ad5295f979a5268ce9433f1c",
    "types": {
      "IndexedStrategy": [
        {
          "name": "index",
          "type": "uint8",
        },
        {
          "name": "params",
          "type": "bytes",
        },
      ],
      "Vote": [
        {
          "name": "space",
          "type": "address",
        },
        {
          "name": "voter",
          "type": "address",
        },
        {
          "name": "proposalId",
          "type": "uint256",
        },
        {
          "name": "choice",
          "type": "uint8",
        },
        {
          "name": "userVotingStrategies",
          "type": "IndexedStrategy[]",
        },
        {
          "name": "voteMetadataURI",
          "type": "string",
        },
      ],
    },
  },
}
`;
